src/main/java/util/RefreshTimer.java

```diff
@@ -0,0 +1,122 @@
+     *                           Pre-condition: >= 0
+     * @param apiCallsUsedInPreviousRefresh The amount of api used in the last api refresh.
+     *                                      Pre-condition: >= 0
+     * @param apiQuotaBuffer The amount of api calls that is set aside for manual operations by the user.
+     *                       Pre-condition: >= 0
+     * @param minRefreshPeriod The minimal refresh period that will be used.
+     *                         There are 3 conditions that this will be used.
+     *                         1) During application initialisation, when apiCallsUsedInPreviousRefresh is zero.
+     *                         2) When computed refresh rate is < minRefreshPeriod.
+     *                         3) When remainingTimeInMins = 0. This indicates that refreshes can occur in the
+     *                            nearest time possible.
+     *                            Recommended value : RefreshTimer.DEFAULT_REFRESH_PERIOD_IN_MINS
+     *                         Pre-condition: > 0
+     * @return Returns computed refresh period.
+     */
+    @SuppressWarnings("PMD")
+    public static long computeRefreshTimerPeriod(int apiQuota, long remainingTimeInMins,
+                                                 int apiCallsUsedInPreviousRefresh,
+                                                 int apiQuotaBuffer, int minRefreshPeriod) {
+
+        assert apiQuota >= 0 && remainingTimeInMins >= 0 && apiCallsUsedInPreviousRefresh >= 0
+                && minRefreshPeriod > 0 && apiQuotaBuffer >= 0;
+
+        final int bufferTime = 1;
+
+        if ((apiQuota > apiQuotaBuffer && apiCallsUsedInPreviousRefresh == 0) || remainingTimeInMins == 0) {
+            return minRefreshPeriod;
+        }
+
+        long refreshTimeInMins;
+
+        if (isQuotaInsufficient(apiQuota, apiCallsUsedInPreviousRefresh)) {
+            refreshTimeInMins = remainingTimeInMins + bufferTime;
+            return Math.max(refreshTimeInMins, minRefreshPeriod);
+        }
+
+        int remainingApiQuota = apiQuota - apiQuotaBuffer;
+        int noOfRefreshAllowed = remainingApiQuota / apiCallsUsedInPreviousRefresh;
+        refreshTimeInMins = (long) Math.ceil(remainingTimeInMins / (double) noOfRefreshAllowed);
+
+        if (refreshTimeInMins == remainingTimeInMins) {
+            refreshTimeInMins = refreshTimeInMins + bufferTime;
+        }
+
+        return Math.max(refreshTimeInMins, minRefreshPeriod);
+    }
+
+    /**
+     * Changes the timer refresh period.
+     * @param periodInMins The period of the TickingTimer in minutes.
+     */
+    public void changeRefreshPeriod(int periodInMins) {
+        this.changePeriod((int) Utility.minsToSecs(periodInMins));
+    }
+
+    private static boolean isQuotaInsufficient(int apiQuota, int apiCallsUsedInPreviousRefresh) {
+        boolean isBelowApiQuotaBufferAllowance = apiQuota <= API_QUOTA_BUFFER;
```

Comment:
Should not depend on `API_QUOTA_BUFFER`. It should be a parameter instead because it is a parameter to the `computeRefreshTimerPeriod` which calls this method.